lib/winapi/thread.c

```diff
@@ -1,7 +1,64 @@
+#include <assert.h>
+#include <string.h>
 #include <processthreadsapi.h>
 #include <winbase.h>
+#include <pdclib/_PDCLIB_xbox_tls.h>
 #include <xboxkrnl/xboxkrnl.h>
 
+VOID NTAPI WinapiThreadStartup (PKSTART_ROUTINE StartRoutine, PVOID StartContext)
+{
+    // Make sure the TLS pointer is valid
+    assert(KeGetCurrentThread()->TlsData);
+
+    // Retrieve TLS area address and point first entry to itself
+    DWORD *TlsData;
+    TlsData = ((DWORD *)KeGetCurrentThread()->TlsData) + 1;
+    TlsData[-1] = (DWORD)TlsData;
+
+    // 16 byte alignment required
+    assert(((DWORD)TlsData & 15) == 0);
+
+    // Copy initial values of TLS area
+    DWORD TlsDataSize;
+    TlsDataSize = _tls_used.EndAddressOfRawData - _tls_used.StartAddressOfRawData;
+    memcpy(TlsData, (void *)_tls_used.StartAddressOfRawData, TlsDataSize);
+
+    // Zero-initialize the rest
+    RtlZeroMemory((char *)TlsData + TlsDataSize, _tls_used.SizeOfZeroFill);
+
+    int res;
+    res = (*(LPTHREAD_START_ROUTINE)StartRoutine)(StartContext);
+    PsTerminateSystemThread(res);
+}
+
+HANDLE CreateThread (LPSECURITY_ATTRIBUTES lpThreadAttributes, SIZE_T dwStackSize, LPTHREAD_START_ROUTINE lpStartAddress, LPVOID lpParameter, DWORD dwCreationFlags, LPDWORD lpThreadId)
+{
+    NTSTATUS status;
+    HANDLE handle;
+
+    if (dwStackSize == 0) {
+        // FIXME: We're directly reading the XBE StackCommit field here.
+        //        A cleaner way with proper structs would be nice.
+        dwStackSize = *((SIZE_T *)0x00010130);
+    }
+
+    ULONG tlssize;
+    // Sum up the required amount of memory, round it up to a multiple of
+    // 16 bytes and add 4 bytes for the self-reference
+    tlssize = (_tls_used.EndAddressOfRawData - _tls_used.StartAddressOfRawData) + _tls_used.SizeOfZeroFill;
+    tlssize = (tlssize + 15) & ~15;
+    tlssize += 4;
+
+    status = PsCreateSystemThreadEx(&handle, 0, dwStackSize, tlssize, (PHANDLE)lpThreadId, (PKSTART_ROUTINE)lpStartAddress, lpParameter, dwCreationFlags & CREATE_SUSPENDED ? TRUE : FALSE, FALSE, WinapiThreadStartup);
```

Comment:
`dwCreationFlags` is never checked for `STACK_SIZE_PARAM_IS_A_RESERVATION`

Reply:
Right, but that's because we don't support that - the kernel always expects the whole stack to be committed memory, growing is not supported (that's why we simply check for problems in `chkstk` instead of probing the memory).

Reply:
We should probably add a comment in the functions which receive this parameter, that it isn't supported (or where you'd expect a declaration to be, so people don't just declare it themselves).

Reply:
Actually, declaring (and using) `STACK_SIZE_PARAM_IS_A_RESERVATION` is fine (I just now added the define for compatibility). It will simply fall back to using committed memory instead.

I added a comment before the declaration of `CreateThread` though to make people aware that the flag is ignored.